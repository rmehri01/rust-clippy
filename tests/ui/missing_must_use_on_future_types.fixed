//@aux-build:proc_macros.rs

#![warn(clippy::missing_must_use_on_future_types)]
#![allow(path_statements, unused, clippy::no_effect)]

extern crate proc_macros;
use proc_macros::{external, with_span};

use std::fmt::Display;
use std::future::Future;
use std::marker::PhantomData;
use std::pin::Pin;
use std::task::{Context, Poll};

#[must_use]
struct BasicStruct; //~ ERROR: missing a `#[must_use]` attribute

impl Future for BasicStruct {
    type Output = ();

    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
        Poll::Ready(())
    }
}

#[must_use]
enum BasicEnum {
    //~^ ERROR: missing a `#[must_use]` attribute
    Var1,
    Var2,
}

impl Future for BasicEnum {
    type Output = ();

    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
        Poll::Ready(())
    }
}

struct NonFuture;

#[must_use]
struct SometimesFuture<T>(PhantomData<T>); //~ ERROR: missing a `#[must_use]` attribute

impl<T: Copy> Future for SometimesFuture<T> {
    type Output = ();

    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
        Poll::Ready(())
    }
}

#[derive(Debug)]
#[must_use]
pub struct DeriveStruct; //~ ERROR: missing a `#[must_use]` attribute

impl Future for DeriveStruct {
    type Output = ();

    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
        Poll::Ready(())
    }
}

#[must_use]
struct AlreadyMustUse;

impl Future for AlreadyMustUse {
    type Output = ();

    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
        Poll::Ready(())
    }
}

external! {
    struct IgnoreExternal;

    impl Future for IgnoreExternal {
        type Output = ();

        fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
            Poll::Ready(())
        }
    }
}

with_span! { span
    struct LocalMacro;

    impl Future for LocalMacro {
        type Output = ();

        fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
            Poll::Ready(())
        }
    }
}
