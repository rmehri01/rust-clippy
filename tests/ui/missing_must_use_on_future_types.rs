#![warn(clippy::missing_must_use_on_future_types)]
#![allow(path_statements, unused, clippy::no_effect)]

use std::fmt::Display;
use std::future::Future;
use std::marker::PhantomData;
use std::pin::Pin;
use std::task::{Context, Poll};

// basic struct case
struct BasicStruct;

impl Future for BasicStruct {
    type Output = ();

    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
        Poll::Ready(())
    }
}

// basic enum case
enum BasicEnum {
    Var1,
    Var2,
}

impl Future for BasicEnum {
    type Output = ();

    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
        Poll::Ready(())
    }
}

// should be ignored if type doesn't implement `Future`
struct NonFuture;

// should still trigger if a type only sometimes implements `Future`
struct SometimesFuture<T>(PhantomData<T>);

impl<T: Copy> Future for SometimesFuture<T> {
    type Output = ();

    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
        Poll::Ready(())
    }
}

// should be ignored on trait objects
trait Trait {}

impl Future for dyn Trait {
    type Output = ();

    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
        Poll::Ready(())
    }
}

struct TraitImpl {}
impl Trait for TraitImpl {}

fn trait_obj() -> Box<dyn Trait> {
    Box::new(TraitImpl {})
}

// struct with multiple fields and impls
#[derive(Debug)]
pub struct ComplexStruct {
    x: usize,
    y: &'static str,
}

impl ComplexStruct {
    fn sum(&self) -> usize {
        self.x + self.y.len()
    }
}

impl Display for ComplexStruct {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "complex")
    }
}

impl Future for ComplexStruct {
    type Output = ();

    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
        Poll::Ready(())
    }
}

// should be ignored on already #[must_use] struct
#[must_use]
struct AlreadyMustUse;

impl Future for AlreadyMustUse {
    type Output = ();

    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
        Poll::Ready(())
    }
}

fn main() {
    BasicStruct;
    BasicEnum::Var2;
    NonFuture;
    SometimesFuture::<String>(PhantomData);
    trait_obj();
    ComplexStruct { x: 42, y: "complex" };
    AlreadyMustUse;
}
