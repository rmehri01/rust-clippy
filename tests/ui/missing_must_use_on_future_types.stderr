error: this struct implements `Future` but is missing a `#[must_use]` attribute
  --> tests/ui/missing_must_use_on_future_types.rs:11:1
   |
LL | struct BasicStruct;
   | ^^^^^^^^^^^^^^^^^^
   |
   = note: `-D clippy::missing-must-use-on-future-types` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::missing_must_use_on_future_types)]`
help: add the attribute
   |
LL + #[must_use]
LL | struct BasicStruct;
   |

error: this struct implements `Future` but is missing a `#[must_use]` attribute
  --> tests/ui/missing_must_use_on_future_types.rs:22:1
   |
LL | enum BasicEnum {
   | ^^^^^^^^^^^^^^
   |
help: add the attribute
   |
LL + #[must_use]
LL | enum BasicEnum {
   |

error: this struct implements `Future` but is missing a `#[must_use]` attribute
  --> tests/ui/missing_must_use_on_future_types.rs:39:1
   |
LL | struct SometimesFuture<T>(PhantomData<T>);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: add the attribute
   |
LL + #[must_use]
LL | struct SometimesFuture<T>(PhantomData<T>);
   |

error: this struct implements `Future` but is missing a `#[must_use]` attribute
  --> tests/ui/missing_must_use_on_future_types.rs:69:1
   |
LL | pub struct ComplexStruct {
   | ^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: add the attribute
   |
LL + #[must_use]
LL | pub struct ComplexStruct {
   |

error: aborting due to 4 previous errors

